### ISSUE TITLE
Is it a trick to speed up the training to add the features of all nodes to the GPU before training?

### ISSUE THREAD

**Issue by maqy1995** [dmlc/dgl#1697]
## ❓ Questions and Help

When running the GraphSAGE example(the code I use is 'dgl/examples/pytorch/graphsage/train_sampling.py '), I found that passing the features of all nodes into the GPU before starting training can improve the training speed, like this:

```
    # passing the features of all nodes into the GPU before starting training
    g.ndata['features'] = g.ndata['features'].to(device)
    # line 180
    # Training loop
    avg = 0
    iter_tput = []
    for epoch in range(args.num_epochs):
        tic = time.time()

        # Loop over the dataloader to sample the computation dependency graph as a list of
        # blocks.
        for step, blocks in enumerate(dataloader):
            tic_step = time.time()
        .......
        .......
```
    
  This can reduce the epoch training time from 20s to about 11s, at the same time, the increased gpu memory uses about 1G. 
  This seems to be very cost-effective. Are there other hidden dangers of this method?

**Comment by mufeili** [comment]
To move all features to GPU at the beginning avoids repeating it for each mini-batch, which can be time-consuming. In general this should be preferred if your GPU can hold all node features simultaneously.

**Comment by maqy1995** [comment]
Thanks for your reply, I will close this issue.

**Comment by cfangplus** [comment]
I think this could be created as a feature, for instance, we could provide an option to user and it's up to the user to decide whether to move features to GPU before training or not. @mufeili

**Comment by cfangplus** [comment]
I also have a question, we run the code 'dgl/examples/pytorch/graphsage/train_sampling.py' and after only two epoches, the loss seems to have convergented and accuracy reached around 95%, I am really confused, could you help to explain that? @mufeili

**Comment by mufeili** [comment]
> I think this could be created as a feature, for instance, we could provide an option to user and it's up to the user to decide whether to move features to GPU before training or not. @mufeili

I think it depends on if GPU memory can be a bottleneck here. For the particular example we are talking about, can you check the amount of GPU memory occupied? If that's not very large, then I don't think it will be an issue for most users.

**Comment by mufeili** [comment]
> I also have a question, we run the code 'dgl/examples/pytorch/graphsage/train_sampling.py' and after only two epoches, the loss seems to have convergented and accuracy reached around 95%, I am really confused, could you help to explain that? @mufeili

Reddit is an easy dataset to tackle. With sampling-based training, we have a lot more iterations with two epochs than the case of full graph training.

**Comment by maqy1995** [comment]
> > I think this could be created as a feature, for instance, we could provide an option to user and it's up to the user to decide whether to move features to GPU before training or not. @mufeili
> 
> I think it depends on if GPU memory can be a bottleneck here. For the particular example we are talking about, can you check the amount of GPU memory occupied? If that's not very large, then I don't think it will be an issue for most users.

I tested this with 'dgl/examples/pytorch/graphsage/train_sampling.py' in the master branch, the only difference in the test code is that the following code () was added before training(line 187):
```
g.ndata['features'] = g.ndata['features'].to(device)
```
The test results are as follows:
Original code：

>>Epoch 00005 | Step 00000 | Loss 0.4230 | Train Acc 0.8980 | Speed (samples/sec) 15989.7564 | GPU 518.5 MiB
Epoch 00005 | Step 00020 | Loss 0.5180 | Train Acc 0.8950 | Speed (samples/sec) 15998.8109 | GPU 518.5 MiB
Epoch 00005 | Step 00040 | Loss 0.4800 | Train Acc 0.8840 | Speed (samples/sec) 16006.2948 | GPU 518.5 MiB
Epoch 00005 | Step 00060 | Loss 0.4600 | Train Acc 0.8870 | Speed (samples/sec) 16013.0906 | GPU 518.5 MiB
Epoch 00005 | Step 00080 | Loss 0.4784 | Train Acc 0.8900 | Speed (samples/sec) 16020.2303 | GPU 518.5 MiB
Epoch 00005 | Step 00100 | Loss 0.4365 | Train Acc 0.8930 | Speed (samples/sec) 16027.0289 | GPU 518.5 MiB
Epoch 00005 | Step 00120 | Loss 0.4927 | Train Acc 0.8990 | Speed (samples/sec) 16033.7174 | GPU 518.5 MiB
Epoch 00005 | Step 00140 | Loss 0.4505 | Train Acc 0.9040 | Speed (samples/sec) 16038.4041 | GPU 518.5 MiB
Epoch Time(s): 13.2684
100%|██████████| 233/233 [00:26<00:00,  8.67it/s]
100%|██████████| 233/233 [00:11<00:00, 20.90it/s]
......
Epoch 00019 | Step 00000 | Loss 0.3978 | Train Acc 0.8940 | Speed (samples/sec) 16031.0585 | GPU 834.5 MiB
Epoch 00019 | Step 00020 | Loss 0.3461 | Train Acc 0.9190 | Speed (samples/sec) 16030.8018 | GPU 834.5 MiB
Epoch 00019 | Step 00040 | Loss 0.4152 | Train Acc 0.8880 | Speed (samples/sec) 16030.6452 | GPU 834.5 MiB
Epoch 00019 | Step 00060 | Loss 0.4075 | Train Acc 0.9020 | Speed (samples/sec) 16029.7481 | GPU 834.5 MiB
Epoch 00019 | Step 00080 | Loss 0.3560 | Train Acc 0.9160 | Speed (samples/sec) 16028.5475 | GPU 834.5 MiB
Epoch 00019 | Step 00100 | Loss 0.3508 | Train Acc 0.9120 | Speed (samples/sec) 16029.0355 | GPU 834.5 MiB
Epoch 00019 | Step 00120 | Loss 0.4248 | Train Acc 0.9140 | Speed (samples/sec) 16028.9887 | GPU 834.5 MiB
Epoch 00019 | Step 00140 | Loss 0.3839 | Train Acc 0.8980 | Speed (samples/sec) 16028.2940 | GPU 834.5 MiB
Epoch Time(s): 13.8294
Eval Acc 0.9455
Avg epoch time: 13.657368771235149


After adding feature to GPU:

>>Epoch 00005 | Step 00000 | Loss 0.5363 | Train Acc 0.8840 | Speed (samples/sec) 81503.7917 | GPU 1078.9 MiB
Epoch 00005 | Step 00020 | Loss 0.3840 | Train Acc 0.9060 | Speed (samples/sec) 81400.7925 | GPU 1078.9 MiB
Epoch 00005 | Step 00040 | Loss 0.4295 | Train Acc 0.8950 | Speed (samples/sec) 81363.3032 | GPU 1078.9 MiB
Epoch 00005 | Step 00060 | Loss 0.4291 | Train Acc 0.8960 | Speed (samples/sec) 81372.6457 | GPU 1078.9 MiB
Epoch 00005 | Step 00080 | Loss 0.4384 | Train Acc 0.9020 | Speed (samples/sec) 81373.5202 | GPU 1078.9 MiB
Epoch 00005 | Step 00100 | Loss 0.4167 | Train Acc 0.9060 | Speed (samples/sec) 81395.9015 | GPU 1078.9 MiB
Epoch 00005 | Step 00120 | Loss 0.4074 | Train Acc 0.9070 | Speed (samples/sec) 81393.7110 | GPU 1078.9 MiB
Epoch 00005 | Step 00140 | Loss 0.5116 | Train Acc 0.8780 | Speed (samples/sec) 81391.4800 | GPU 1078.9 MiB
Epoch Time(s): 5.8531
100%|██████████| 233/233 [00:11<00:00, 20.84it/s]
100%|██████████| 233/233 [00:11<00:00, 20.87it/s]
Eval Acc 0.9461
......
Epoch 00019 | Step 00000 | Loss 0.3833 | Train Acc 0.9060 | Speed (samples/sec) 78900.2102 | GPU 1396.8 MiB
Epoch 00019 | Step 00020 | Loss 0.4250 | Train Acc 0.9060 | Speed (samples/sec) 78884.3071 | GPU 1396.8 MiB
Epoch 00019 | Step 00040 | Loss 0.3411 | Train Acc 0.9100 | Speed (samples/sec) 78854.2983 | GPU 1396.8 MiB
Epoch 00019 | Step 00060 | Loss 0.4407 | Train Acc 0.9020 | Speed (samples/sec) 78853.0732 | GPU 1396.8 MiB
Epoch 00019 | Step 00080 | Loss 0.3890 | Train Acc 0.9070 | Speed (samples/sec) 78830.1895 | GPU 1396.8 MiB
Epoch 00019 | Step 00100 | Loss 0.3872 | Train Acc 0.9100 | Speed (samples/sec) 78812.8802 | GPU 1396.8 MiB
Epoch 00019 | Step 00120 | Loss 0.3479 | Train Acc 0.9170 | Speed (samples/sec) 78828.9618 | GPU 1396.8 MiB
Epoch 00019 | Step 00140 | Loss 0.3224 | Train Acc 0.9180 | Speed (samples/sec) 78823.7870 | GPU 1396.8 MiB
Epoch Time(s): 6.2348
Avg epoch time: 6.149972502390543

Test results show that training speed and inference speed(layer 0 from 25s to 11s) have been improved, and about 600MB of gpu memory usage increased in this way.

PS: The environment is: cpu:3700x,  gpu:2060s,  cuda10.2, pytorch1.5.0, dgl-cu102 0.4.3.post2

**Comment by mufeili** [comment]
Copy @BarclayII for visibility, probably we can add a flag there for copying all node features to GPU at the beginning and briefly mention that in README?

**Comment by BarclayII** [comment]
I'm a bit hesitant on this.  Such trick is not always applicable to all datasets.  In fact, if the dataset goes bigger to the scale like OGB ones, then one would still have to put the node features on CPU and copy the necessary features of the minibatch to GPU on demand.

**Comment by cfangplus** [comment]
@BarclayII yea, that's why we suggest to provide an option to user so that they could decide whether to copy or not.
